docs/design/Path concatenation.md

```diff
@@ -0,0 +1,94 @@
+## Paths in DF2
+
+It is expected that the use of paths in DF2 (similar in nature to Unix file paths, but not exactly the same) will be extensive, as will be concatention operations on them: `GetFullPath` and `GetRelativePath`. This document will provide a specification for how these operations are to behave as well as contain some considerations that will drive the design of concatenation functions.
+
+## Definitions
+
+_Char_ is a Unicode code point. There are special _chars_ which have specific meaning in DF2 paths:
+- _Separator_ is a _char_ corresponding to the code point `U+002F` - `Slash`.
+- _Dot_ is a _char_ corresponding to the code point `U+002E` - `Full stop`.
+
+Two _chars_ are equal if they refer to the same code point.
+
+[Rationale: to support the performance of concatenation functions, it is neccessary to impose the restriction that equality comparisons be strictly ordinal.]
+
+_Segments_ are sequences of _chars_ that have no _separators_ in them. There are special _segments_ which have specific meaning in DF2 paths:
+- _Skip segment_ is a _segment_ that consists of one _dot_ or has no _chars_ at all.
+- _Backtrack segment_ is a _segment_ that consists of two _dots_.
+
+Two _segments_ are equal if all their _chars_ are equal, except for _skip segments_, which are always equal to each other.
+
+DF2 _paths_ are sequences of _segments_ and _separators_. It is illegal for two _segments_ to follow each other. Two _separators_ that have no _chars_ between them are treated as a _separator_-_skip segment_-_separator_ triplet.
+
+Three types of _paths_ represent subsets of the above definition:
+- _Normalized paths_ have no _skip segments_ in them and always have the same number of _separators_ as _segments_. _Normalized paths_ that have at least one _segment_ always start with a _separator_ and end with a _segment_.
+
+[Rationale: it is expected that most implementations will choose to normalize the paths as they enter the process, and work with normalized paths in memory due their superior performance characteristics. The restriction on the start and end of the normalized path are to ensure they always have the same number of _segments_ as they have _separators_, which simplifies the implementation of concatenation functions.]
+
+- _Absolute paths_ are _normalized paths_ that have no _backtrack segments_.
+- _Relative paths_ are _normalized paths_ that can only have _backtrack segments_ at the beginning of the _path_.
```

Comment:
I think this could refer to just paths. What is described here would then become a `normalized absolute/relative Path`.

Reply:
FWIW, I think it is worth keeping the concept of a normalized path because it makes the normalization procedure have "one kind" of an output - a normalized path.

It is worth noting that normalization as written will always produce a valid relative path, and an absolute path is always a valid relative path, so, technically, relative paths _are_ normalized paths. I just wrote it that way because it is clearer, in my opinion, to have this intermediate concept that unifies them.

Let me know if you think that's enough of a justification to keep it or not.

Reply:
yep! makes sense. Just wasn't obvious to me that absolute -> relative -> normalized.